crates/sequencer/src/sequencer.rs

```diff
@@ -999,3 +1048,88 @@ where
         Ok(updates)
     }
 }
+
+fn get_l1_fee_rate_range<C, Da>(
+    storage: C::Storage,
+    rule_enforcer: SoftConfirmationRuleEnforcer<C, Da::Spec>,
+) -> Result<RangeInclusive<u128>, anyhow::Error>
+where
+    C: Context,
+    Da: DaService,
+{
+    let mut working_set = WorkingSet::<C>::new(storage);
+
+    rule_enforcer
+        .get_next_min_max_l1_fee_rate(&mut working_set)
+        .map_err(|e| anyhow::anyhow!("Error reading min max l1 fee rate: {}", e))
+}
+
+async fn da_block_monitor<C, Da>(
+    da_service: Da,
+    storage: C::Storage,
+    rule_enforcer: SoftConfirmationRuleEnforcer<C, Da::Spec>,
+    sender: mpsc::Sender<L1Data<Da>>,
+    loop_interval: u64,
+) where
+    C: Context,
+    Da: DaService + Clone,
+{
+    loop {
+        let l1_fee_rate_range =
+            match get_l1_fee_rate_range::<C, Da>(storage.clone(), rule_enforcer.clone()) {
+                Ok(fee_rate_range) => fee_rate_range,
+                Err(e) => {
+                    error!("Could not fetch L1 fee rate range: {}", e);
+                    continue;
+                }
+            };
+        let l1_data = match get_da_block_data(da_service.clone(), l1_fee_rate_range).await {
+            Ok(l1_data) => l1_data,
+            Err(e) => {
+                error!("Could not fetch L1 data, {}", e);
+                continue;
+            }
+        };
+
+        let _ = sender.send(l1_data).await;
+
+        sleep(Duration::from_millis(loop_interval)).await;
+    }
+}
+
+async fn get_da_block_data<Da>(
+    da_service: Da,
+    l1_fee_rate_range: RangeInclusive<u128>,
+) -> anyhow::Result<L1Data<Da>>
+where
+    Da: DaService,
+{
+    let last_finalized_height = match da_service.get_last_finalized_block_header().await {
+        Ok(header) => header.height(),
+        Err(e) => {
+            return Err(anyhow!("Finalized height: {}", e));
+        }
+    };
+
+    let last_finalized_block = match da_service.get_block_at(last_finalized_height).await {
+        Ok(block) => block,
+        Err(e) => {
+            return Err(anyhow!("Finalized block: {}", e));
+        }
+    };
+
+    debug!(
+        "Sequencer: last finalized height: {:?}",
+        last_finalized_height
+    );
+
+    let l1_fee_rate = match da_service.get_fee_rate().await {
+        Ok(fee_rate) => fee_rate,
+        Err(e) => {
+            return Err(anyhow!("L1 fee rate: {}", e));
+        }
+    };
+    let l1_fee_rate = l1_fee_rate.clamp(*l1_fee_rate_range.start(), *l1_fee_rate_range.end());
```

Comment:
i think it's a better idea to clamp l1 fee rate on the block producftion level rather than where we pull da data.

it will give the block production loop more control over the l1 fee rate.

in the future we might want to decide on things according to l1 fee rate.